data/code/lib.c

```diff
@@ -323,3 +324,170 @@ int itemshufflemgr_calcslot_ex(const unsigned int* const itemshufflemgr,
 }
 
 #endif
+
+#if SECTIONED_COURSES
+
+static unsigned short section_count = 0;
+
+// Due to the nature of the compiler, a lot of the code had to be rewritten in ASM
+// so that the compiler would not ignore it, and thus break this code patch.
+// To compensate, nearly every set of ASM instructions has a description of what it's doing.
+
+// Override the lap count in a section course to be the number of section points.
+void section_lap_override()
+{
+    register unsigned short reg9 asm("r9");
+    asm("or %r0,%r3,%r3 \t\n");
+
+    reg9 = section_count;
+
+    if (reg9 != 0)
+    {
+        if (reg9 > 9)
+        {
+            asm("li %r10, 0x09       \t\n"
+                "sth %r10,0x2e(%r31) \t\n");
+        }
+        else
+        {
+            asm("sth %r9, 0x2e(%r31)");
+        }
+    }
+}
+
+// Reset the section counter.
+void reset_sect_count()
+{
+    asm("or %r31, %r3, %r3");  // Run hijacked instruction.
+    section_count = 0;
+}
+
+// Count the section points to jury-rig the "max laps" count to always be the section number.
+void count_sect_points()
+{
+    register unsigned int reg4 asm("r4");
+    register unsigned char reg9 asm("r9");
+
+    asm("stw %r4, 0x8(%r31)                 \t\n");  // Run hijacked instruction.
+
+    reg9 = *(unsigned char*)(reg4 + 0x0018);  // "Shortcut Point" flag.
+
+    if (reg9 != 0)  // Disregard invalid section points.
+        return;
+
+    reg9 = *(unsigned char*)(reg4 + 0x001B);  // "Lap Checkpoint" flag.
+
+    if (reg9 == 0)  // Disregard checkpoints with an unset section flag.
+        return;
+
+    ++section_count;
+}
+
+// ASM rips of lap and goal functions since they don't exist in retail.
+void lap_increment()
+{
+    asm("lwz        %r4,0x2c(%r3) \t\n"
+        "lwz        %r0,0xc(%r3)  \t\n"
+        "cmpw       %r4,%r0       \t\n"
+        "bgelr                    \t\n"
+        "addi       %r0,%r4,0x1   \t\n"
+        "stw        %r0,0x2c(%r3) \t\n");
+}
+
+void do_goal_routine()
```

Comment:
Nit: I always find "do" too vague. Would "start" (`start_goal_routine()`) be sensible?